File: pb/src/main/java/org/sqlbuilder/pb/ResolvingUpdater.java
package org.sqlbuilder.pb;

import org.sqlbuilder.annotations.ProvidesIdTo;
import org.sqlbuilder.common.Progress;
import org.sqlbuilder.common.Update;
import org.sqlbuilder.common.Utils;
import org.sqlbuilder.pb.foreign.FnAlias;
import org.sqlbuilder.pb.foreign.VnAlias;
import org.sqlbuilder.pb.foreign.VnRoleAlias;
import org.sqlbuilder.pb.objects.Word;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.util.Properties;

public class ResolvingUpdater extends ResolvingInserter
{
	public ResolvingUpdater(final Properties conf) throws IOException, ClassNotFoundException
	{
		super(conf);

		// output
		this.outDir = new File(conf.getProperty("pm_outdir_updated", "sql/data_updated"));
		if (!this.outDir.exists())
		{
			//noinspection ResultOfMethodCallIgnored
			this.outDir.mkdirs();
		}
	}

	@Override
	public void insert() throws FileNotFoundException
	{
		try (@ProvidesIdTo(type = Word.class) var ignored30 = Word.COLLECTOR.open())
		{
			insertWords();
			insertFnAliases();
			insertVnAliases();
			insertVnRoleAliases();
		}
	}

	@Override
	protected void insertWords() throws FileNotFoundException
	{
		Progress.tracePending("collector", "word");
		final String wordidCol = names.column("words.wordid");
		final String wordCol = names.column("words.word");
		Update.update(Word.COLLECTOR.keySet(), new File(outDir, names.updateFile("words")), header, names.table("words"), //
				wordResolver, //
				resolved -> String.format("%s=%s", wordidCol, Utils.nullableInt(resolved)), //
				resolving -> String.format("%s='%s'", wordCol, Utils.escape(resolving)));
		Progress.traceDone();
	}

	@Override
	protected void insertFnAliases() throws FileNotFoundException
	{
		Progress.tracePending("set", "fnalias");
		final String fnframeidCol = names.column("pbrolesets_fnframes.fnframeid");
		final String vnclassCol = names.column("pbrolesets_fnframes.fnframe");
		Update.update(FnAlias.SET, new File(outDir, names.updateFile("pbrolesets_fnframes")), header, names.table("pbrolesets_fnframes"), //
				fnFrameResolver, //
				resolved -> String.format("%s=%s", fnframeidCol, Utils.nullableInt(resolved)), //
				resolving -> String.format("%s='%s'", vnclassCol, Utils.escape(resolving)));
		Progress.traceDone();
	}

	@Override
	protected void insertVnAliases() throws FileNotFoundException
	{
		Progress.tracePending("set", "vnalias");
		final String vnclassidCol = names.column("pbrolesets_vnclasses.vnclassid");
		final String vnclassCol = names.column("pbrolesets_vnclasses.vnclass");
		Update.update(VnAlias.SET, new File(outDir, names.updateFile("pbrolesets_vnclasses")), header, names.table("pbrolesets_vnclasses"), //
				vnClassResolver, //
				resolved -> String.format("%s=%s", vnclassidCol, Utils.nullableInt(resolved)), //
				resolving -> String.format("%s='%s'", vnclassCol, Utils.escape(resolving)));
		Progress.traceDone();
	}

	@Override
	protected void insertVnRoleAliases() throws FileNotFoundException
	{
		Progress.tracePending("set", "vnaliasrole");
		final String vnclassidCol = names.column("pbroles_vnroles.vnclassid");
		final String vnroleidCol = names.column("pbroles_vnroles.vnroleid");
		final String vnroletypeidCol = names.column("pbroles_vnroles.vnroletypeid");
		final String vnclassCol = names.column("pbroles_vnroles.vnclass");
		final String vnroleCol = names.column("pbroles_vnroles.vntheta");
		Update.update(VnRoleAlias.SET, new File(outDir, names.updateFile("pbroles_vnroles")), header, names.table("pbroles_vnroles"), //
				vnClassRoleResolver, //
				resolved -> resolved == null ?
						String.format("%s=NULL,%s=NULL,%s=NULL", vnclassidCol, vnroleidCol, vnroletypeidCol) :
						String.format("%s=%s,%s=%s,%s=%s", vnclassidCol, Utils.nullableInt(resolved.first), vnroleidCol, Utils.nullableInt(resolved.second), vnroletypeidCol, Utils.nullableInt(resolved.third)), //
				resolving -> String.format("%s='%s' AND %s='%s'", vnclassCol, Utils.escape(resolving.first), vnroleCol, Utils.escape(resolving.second)));
		Progress.traceDone();
	}
}

File: pb/src/main/java/org/sqlbuilder/pb/ResolvingUpdater.kt
package org.sqlbuilder.pb

import org.sqlbuilder.common.Progress.traceDone
import org.sqlbuilder.common.Progress.traceSaving
import org.sqlbuilder.common.Update.update
import org.sqlbuilder.common.Utils.escape
import org.sqlbuilder.common.Utils.nullableInt
import org.sqlbuilder.pb.foreign.RoleSetToFn
import org.sqlbuilder.pb.foreign.RoleSetToVn
import org.sqlbuilder.pb.foreign.RoleToVn
import org.sqlbuilder.pb.objects.Word
import java.io.File
import java.io.FileNotFoundException
import java.util.*

class ResolvingUpdater(conf: Properties) : ResolvingInserter(conf) {
    init {
        // output
        outDir = File(conf.getProperty("pm_outdir_updated", "sql/data_updated"))
        if (!outDir.exists()) {
            outDir.mkdirs()
        }
    }

    @Throws(FileNotFoundException::class)
    override fun insert() {
        Word.COLLECTOR.open().use {
            insertWords()
            insertFnFrameAliases()
            insertVnClassAliases()
            insertVnRoleAliases()
            insertFnFeAliases()
        }
    }

    @Throws(FileNotFoundException::class)
    override fun insertWords() {
        traceSaving("word")
        val wordidCol = names.column("words.wordid")
        val wordCol = names.column("words.word")
        update(
            Word.COLLECTOR,
            File(outDir, names.updateFile("words")),
            header,
            names.table("words"),
            wordResolver,
            { resolved -> "$wordidCol=${nullableInt(resolved)}" },
            { resolving -> "$wordCol='${escape(resolving)}'" })
        traceDone()
    }

    @Throws(FileNotFoundException::class)
    override fun insertVnClassAliases() {
        traceSaving("vnalias")
        val vnclassidCol = names.column("pbrolesets_vnclasses.vnclassid")
        val vnclassCol = names.column("pbrolesets_vnclasses.vnclass")
        update(
            RoleSetToVn.SET,
            File(outDir, names.updateFile("pbrolesets_vnclasses")),
            header,
            names.table("pbrolesets_vnclasses"),
            vnClassResolver,
            { resolved -> "$vnclassidCol=${nullableInt(resolved)}" },
            { resolving -> "$vnclassCol='${escape(resolving)}'" })
        traceDone()
    }

    @Throws(FileNotFoundException::class)
    override fun insertFnFrameAliases() {
        traceSaving("fnalias")
        val fnframeidCol = names.column("pbrolesets_fnframes.fnframeid")
        val vnclassCol = names.column("pbrolesets_fnframes.fnframe")
        update(
            RoleSetToFn.SET,
            File(outDir, names.updateFile("pbrolesets_fnframes")),
            header,
            names.table("pbrolesets_fnframes"),
            fnFrameResolver,
            { resolved -> "$fnframeidCol=${nullableInt(resolved)}" },
            { resolving -> "$vnclassCol='${escape(resolving)}'" })
        traceDone()
    }

    @Throws(FileNotFoundException::class)
    override fun insertVnRoleAliases() {
        traceSaving("vnaliasrole")
        val vnClassidCol = names.column("pbroles_vnroles.vnclassid")
        val vnRoleidCol = names.column("pbroles_vnroles.vnroleid")
        val vnRoletypeidCol = names.column("pbroles_vnroles.vnroletypeid")
        val vnClassCol = names.column("pbroles_vnroles.vnclass")
        val vnRoleCol = names.column("pbroles_vnroles.vnrole")
        update(
            RoleToVn.SET,
            File(outDir, names.updateFile("pbroles_vnroles")),
            header,
            names.table("pbroles_vnroles"),
            vnRoleResolver,
            { resolved ->
                if (resolved == null)
                    "$vnClassidCol=NULL,$vnRoleidCol=NULL,$vnRoletypeidCol=NULL"
                else
                    "$vnClassidCol=${nullableInt(resolved.first)},$vnRoleidCol=${nullableInt(resolved.second)},$vnRoletypeidCol=${nullableInt(resolved.third)}"
            },
            { resolving -> "$vnClassCol='${escape(resolving.first)}' AND $vnRoleCol='${escape(resolving.second)}'" })
        traceDone()
    }

    @Throws(FileNotFoundException::class)
    override fun insertFnFeAliases() {
        traceSaving("fnaliasrole")
        val fnframeidCol = names.column("pbroles_fnfes.fnframeid")
        val fnfeidCol = names.column("pbroles_fnfes.fnfeid")
        val fnfetypeidCol = names.column("pbroles_fnfes.fnfetypeid")
        val fnframeCol = names.column("pbroles_fnfes.fnframe")
        val fnfeCol = names.column("pbroles_fnfes.fnfe")
        update(
            RoleToVn.SET,
            File(outDir, names.updateFile("pbroles_fnfes")),
            header,
            names.table("pbroles_fnfes"),
            fnFeResolver,
            { resolved ->
                if (resolved == null)
                    "$fnframeidCol=NULL,$fnfeidCol=NULL,$fnfetypeidCol=NULL"
                else
                    "$fnframeidCol=${nullableInt(resolved.first)},$fnfeidCol=${nullableInt(resolved.second)},$fnfetypeidCol=${nullableInt(resolved.third)}"
            },
            { resolving -> "$fnframeCol='${escape(resolving.first)}' AND $fnfeCol='${escape(resolving.second)}'" })
        traceDone()
    }
}
